### ISSUE TITLE
ppc64 IBM i support?

### ISSUE THREAD

**Issue by ThePrez** [python-greenlet/greenlet#145]
Attempts to use `bpython` on IBM i fail with the following:
```
Traceback (most recent call last):
  File "/QOpenSys/pkgs/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/QOpenSys/pkgs/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/QOpenSys/pkgs/lib/python3.6/site-packages/bpython/__main__.py", line 31, in <module>
    sys.exit(main())
  File "/QOpenSys/pkgs/lib/python3.6/site-packages/bpython/curtsies.py", line 191, in main
    exit_value = repl.mainloop(True, paste)
  File "/QOpenSys/pkgs/lib/python3.6/site-packages/bpython/curtsies.py", line 126, in mainloop
    self.process_event_and_paint(e)
  File "/QOpenSys/pkgs/lib/python3.6/site-packages/bpython/curtsies.py", line 90, in process_event_and_paint
    self.process_event(e)
  File "/QOpenSys/pkgs/lib/python3.6/site-packages/bpython/curtsiesfrontend/repl.py", line 603, in process_event
    return self.process_key_event(e)
  File "/QOpenSys/pkgs/lib/python3.6/site-packages/bpython/curtsiesfrontend/repl.py", line 732, in process_key_event
    self.on_enter()
  File "/QOpenSys/pkgs/lib/python3.6/site-packages/bpython/curtsiesfrontend/repl.py", line 817, in on_enter
    self.push(self.current_line, insert_into_history=insert_into_history)
  File "/QOpenSys/pkgs/lib/python3.6/site-packages/bpython/curtsiesfrontend/repl.py", line 1087, in push
    self.run_code_and_maybe_finish()
  File "/QOpenSys/pkgs/lib/python3.6/site-packages/bpython/curtsiesfrontend/repl.py", line 1090, in run_code_and_maybe_finish
    r = self.coderunner.run_code(for_code=for_code)
  File "/QOpenSys/pkgs/lib/python3.6/site-packages/bpython/curtsiesfrontend/coderunner.py", line 138, in run_code
    request = self.code_context.switch()
SystemError: <built-in method switch of greenlet.greenlet object at 0x700000000b713d8> returned NULL without setting an error
```
I suspect this is because there's no proper AIX ppc64 implementation (IBM i uses an AIX runtime and instruction set).

I'm willing to help built/test a solution if needed.

**Comment by kskuhlman** [comment]
This may be as simple as refining the existing platform check in slp_platformselect.h:

`#elif defined(__GNUC__) && defined(_ARCH_PPC) && defined(_AIX)`
`#include "platform/switch_ppc_aix.h" /* gcc on AIX/PowerPC */`

Can't test a patch at the moment.. I'm away from my dev box.

**Comment by kskuhlman** [comment]
Copied AIX platform header to IBM i / PASE.    Was able to do a build_ext standalone but not when doing run_tests.
 
$ uname -a
OS400 v156175 3 7 007800000760

$ python3 setup.py develop
running develop
running egg_info
writing greenlet.egg-info/PKG-INFO
writing dependency_links to greenlet.egg-info/dependency_links.txt
writing top-level names to greenlet.egg-info/top_level.txt
reading manifest file 'greenlet.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'greenlet.egg-info/SOURCES.txt'
running build_ext
building 'greenlet' extension
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -O2 -g -maix64 -O2 -g -maix64 -I/QOpenSys/pkgs/include/python3.6m -c greenlet.c -o build/temp.os400-powerpc64-3.6/greenlet.o
creating build/lib.os400-powerpc64-3.6
/QOpenSys/pkgs/lib/python3.6/config-3.6m/ld_so_aix gcc -pthread -bI:/QOpenSys/pkgs/lib/python3.6/config-3.6m/python.exp -Wl,-blibpath:/QOpenSys/pkgs/lib:/QOpenSys/usr/lib build/temp.os400-powerpc64-3.6/greenlet.o -L/QOpenSys/pkgs/lib -o build/lib.os400-powerpc64-3.6/greenlet.so
Linking /src/greenlet.git/build/lib.os400-powerpc64-3.6/greenlet.so to /src/greenlet.git/greenlet.so
Creating /QOpenSys/pkgs/lib/python3.6/site-packages/greenlet.egg-link (link to .)
Adding greenlet 0.4.15 to easy-install.pth file

Installed /src/greenlet.git
Processing dependencies for greenlet==0.4.15
Finished processing dependencies for greenlet==0.4.15

$ python3 setup.py build_ext
running build_ext
building 'greenlet' extension
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -O2 -g -maix64 -O2 -g -maix64 -I/QOpenSys/pkgs/include/python3.6m -c greenlet.c -o build/temp.os400-powerpc64-3.6/greenlet.o
/QOpenSys/pkgs/lib/python3.6/config-3.6m/ld_so_aix gcc -pthread -bI:/QOpenSys/pkgs/lib/python3.6/config-3.6m/python.exp -Wl,-blibpath:/QOpenSys/pkgs/lib:/QOpenSys/usr/lib build/temp.os400-powerpc64-3.6/greenlet.o -L/QOpenSys/pkgs/lib -o build/lib.os400-powerpc64-3.6/greenlet.so


$ python3 run-tests.py
running build_ext
Linking /src/greenlet.git/build/lib.os400-powerpc64-3.6/greenlet.so to /src/greenlet.git/greenlet.so
python 3.6.8 (64 bit) using greenlet 0.4.15 from /src/greenlet.git/greenlet.so
ld: 0711-317 ERROR: Undefined symbol: vtable for __cxxabiv1::__class_type_info
ld: 0711-317 ERROR: Undefined symbol: __gxx_personality_v0
ld: 0711-317 ERROR: Undefined symbol: .__cxa_begin_catch
ld: 0711-317 ERROR: Undefined symbol: .__cxa_end_catch
ld: 0711-317 ERROR: Undefined symbol: .__cxa_allocate_exception
ld: 0711-317 ERROR: Undefined symbol: .__cxa_throw
ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.
collect2: error: ld returned 8 exit status
error: command '/QOpenSys/pkgs/lib/python3.6/config-3.6m/ld_so_aix' failed with exit status 1

**Comment by ThePrez** [comment]
@kadler, any quick thoughts on that failure?

**Comment by kadler** [comment]
It's giving errors about loading C++ symbols, while the code looks to be all C code, so that's odd. It would be interesting to see what `-bnoquiet` gives. Could try `LDFLAGS='-Wl,-bnoquiet' python3 run-tests.py`. I can't remember if that works to pass additional LDFLAGS through or not.

**Comment by kskuhlman** [comment]
Thanks.. that's more informative:

```
$ LDFLAGS='-Wl,-bnoquiet' python3 run-tests.py
running build_ext
Linking /src/greenlet.git/build/lib.os400-powerpc64-3.6/greenlet.so to /src/greenlet.git/greenlet.so
python 3.6.8 (64 bit) using greenlet 0.4.15 from /src/greenlet.git/greenlet.so
(ld): setopt 64
(ld): halt 4
(ld): setopt r/o->w
(ld): setopt rtl
(ld): setopt nortllib
(ld): setopt symbolic:1
(ld): setfflag 4
(ld): savename build/tests/lib.os400-powerpc64-3.6/_test_extension_cpp.so
(ld): filelist 10 3
(ld): setopt noprogram
(ld): entry PyInit__test_extension_cpp
ENTRY: Entry point set to PyInit__test_extension_cpp
(ld): i /lib/crt0_64.o
(ld): i /QOpenSys/pkgs/bin/../lib/gcc/powerpc-ibm-aix6.1.0.0/6.3.0/pthread/ppc64/crtcxa.o
(ld): i /QOpenSys/pkgs/bin/../lib/gcc/powerpc-ibm-aix6.1.0.0/6.3.0/pthread/ppc64/crtdbase.o
(ld): i build/tests/temp.os400-powerpc64-3.6/tests/_test_extension_cpp.o
(ld): lib /QOpenSys/pkgs/lib/libstdc++.a
(ld): lib /usr/lib/libm.a
(ld): lib /QOpenSys/pkgs/lib/libgcc_s.so
(ld): i /QOpenSys/pkgs/bin/../lib/gcc/powerpc-ibm-aix6.1.0.0/6.3.0/pthread/ppc64/libgcc.a
(ld): lib /usr/lib/libpthreads.a
(ld): lib /usr/lib/libc.a
LIBRARY: Symbols imported from import file /QOpenSys/pkgs/lib/libgcc_s.so[shr_64.imp]: 150
LIBRARY: Shared object libpthreads.a[shr_xpg5_64.o]: 346 symbols imported.
LIBRARY: Shared object libc.a[shr_64.o]: 3143 symbols imported.
LIBRARY: Shared object libc.a[posix_aio_64.o]: 20 symbols imported.
LIBRARY: Shared object libc.a[aio_64.o]: 18 symbols imported.
LIBRARY: Shared object libc.a[pse_64.o]: 8 symbols imported.
LIBRARY: Shared object libc.a[dl_64.o]: 4 symbols imported.
LIBRARY: Shared object libc.a[pty_64.o]: 1 symbols imported.
LIBRARY: Shared object libc.a[cthread_64.o]: 25 symbols imported.
LIBRARY: Shared object libc.a[uchar_64.o]: 4 symbols imported.
FILELIST: Number of previously inserted files processed: 10
(ld): imports /QOpenSys/pkgs/lib/python3.6/config-3.6m/python.exp
IMPORTS: Symbols imported from import file /QOpenSys/pkgs/lib/python3.6/config-3.6m/python.exp: 1735
(ld): exports _test_extension_cpp.exp
EXPORTS: Symbols exported: 4
(ld): resolve
RESOLVE: 43 of 8561 symbols were kept.
(ld): addgl /usr/lib/glink64.o
ADDGL: Glink code added for 10 symbols.
(ld): er full
ld: 0711-318 ERROR: Undefined symbols were found.
        The following symbols are in error:
 Symbol                    Inpndx  TY CL Source-File(Object-File) OR Import-File{Shared-object}
                              RLD: Address  Section  Rld-type Referencing Symbol
 ----------------------------------------------------------------------------------------------
 vtable for __cxxabiv1::__class_type_info [30]    ER PR tests/_test_extension_cpp.cpp(build/tests/temp.os400-powerpc64-3.6/tests/_test_extension_cpp.o)
                                   00000848 .data    R_POS    [870]   typeinfo for exception_t
 __gxx_personality_v0      [34]    ER PR tests/_test_extension_cpp.cpp(build/tests/temp.os400-powerpc64-3.6/tests/_test_extension_cpp.o)
                                   000007e0 .data    R_POS    [856]   <_testextensioncpp.rw_>
 .__cxa_begin_catch        [10]    ER PR tests/_test_extension_cpp.cpp(build/tests/temp.os400-powerpc64-3.6/tests/_test_extension_cpp.o)
                                   00000180 .text    R_RBR    [783]   <.test_exception_switch_recurse(int, int)>
                                   000001bc .text    R_RBR    [783]   <.test_exception_switch_recurse(int, int)>
 .__cxa_end_catch          [12]    ER PR tests/_test_extension_cpp.cpp(build/tests/temp.os400-powerpc64-3.6/tests/_test_extension_cpp.o)
                                   000001b0 .text    R_RBR    [783]   <.test_exception_switch_recurse(int, int)>
                                   000001dc .text    R_RBR    [783]   <.test_exception_switch_recurse(int, int)>
                                   000001f0 .text    R_RBR    [783]   <.test_exception_switch_recurse(int, int)>
                                   0000021c .text    R_RBR    [783]   <.test_exception_switch_recurse(int, int)>
 .__cxa_allocate_exception [20]    ER PR tests/_test_extension_cpp.cpp(build/tests/temp.os400-powerpc64-3.6/tests/_test_extension_cpp.o)
                                   00000288 .text    R_RBR    [806]   <.test_exception_throw(int)>
 .__cxa_throw              [22]    ER PR tests/_test_extension_cpp.cpp(build/tests/temp.os400-powerpc64-3.6/tests/_test_extension_cpp.o)
                                   0000029c .text    R_RBR    [806]   <.test_exception_throw(int)>
ER: The return code is 8.
ld: 0711-317 ERROR: Undefined symbol: vtable for __cxxabiv1::__class_type_info
ld: 0711-317 ERROR: Undefined symbol: __gxx_personality_v0
ld: 0711-317 ERROR: Undefined symbol: .__cxa_begin_catch
ld: 0711-317 ERROR: Undefined symbol: .__cxa_end_catch
ld: 0711-317 ERROR: Undefined symbol: .__cxa_allocate_exception
ld: 0711-317 ERROR: Undefined symbol: .__cxa_throw
collect2: error: ld returned 8 exit status
error: command '/QOpenSys/pkgs/lib/python3.6/config-3.6m/ld_so_aix' failed with exit status 1
```

**Comment by kadler** [comment]
Ok, looks like some tests are written in C++.  I didn't notice it in the swath of headers labeled as "C++" by github [here](https://github.com/python-greenlet/greenlet/search?l=c%2B%2B)

Looking at the linker output, it curiously found libstdc++.a instead of libstdc++.so. This is only provided for compatibility with the GCC compilers we get from AIX Toolbox and they are marked "load only" and as you might see, there were no symbols imported from it.

I believe instaling libstdcplusplus-devel should fix this problem.

**Comment by jamadden** [comment]
Fixed in #197
